### DISCUSSION TITLE
RFC: Associated Types

### DISCUSSION THREAD

**Discussion by eudoxia0** [austral/austral#417] [Ideas]
Associated types are an extension to type classes that lets you do things you can't do with Austral's current type class system.

# Example

The simplest example is a type class for representing set-like collections:

```
typeclass Set(T: Type) is
    method member(e: E, set: T): Bool;
end;
```

We can't write this, because `E` is not defined.

Adding support for multi-parameter typeclasses would allow us to write:

```
typeclass Set(T: Type, E: Type) is
    method member(e: E, set: T): Bool;
end;
```

But this is too unconstrained, since we could write an instance like this:

```
generic [U: Type]
instance Set(HashSet[U], Int32) is
    method member(e: Int32, set: HashSet[U]): Bool;
end;
```

Which doesn't make sense.

Associated types solve this problem. An associated type is something like a type alias that is local to the typeclass, where instances have to specify what type to pass for that alias. So, for example, we could write:

```
typeclass Set(T: Type) is
    associated type E: Type;

    method member(e: E, set: T): Bool;
end;
```

Then, instances of `Set` would have to specify what `E` is:

```
generic [U: Type]
instance Set(HashSet[U]) is
    associated type Elem := U;

    method member(e: Elem, set: HashSet[U]): Bool;
end;
```

# Implementation

Relatively involved. Type class resolution is largely untouched, but the parsing and storage of typeclass instances would have to change a lot.

**Comment by L-as** [comment]
What would the syntax for referring to them from outside the class/instance definition look like?

**Reply by eudoxia0** [reply]
How do you mean? Like in a call?

If you were doing:

```
let s: HashSet[Int32] := makeEmptyHashSet();
member(13, set);
```

The compiler would find the instance that satisfies this call, the one for `HashSet[U]`, it would find that `U = Int32`, and it would rewrite `E = Int32`.

The associated type is going to be either just a type parameter or a type specifier from types defined outside the instance, so you'd refer to it as any normal type.

A possibly more useful example:

```
typeclass Foo(T: Type) is
    associated type U: Type;

    method foo(t: T): U;
end;

instance Foo(Float64) is
    associated type U: Nat64;

    method foo(t: Float64): Nat64 is
        -- unimportant
    end;
end;
```

Here you'd call this like this:

```
let n: Nat64 := foo(3.14);
```

In other words: you don't have to refer to a type `U`, you juts refer to whatever the instance makes that type equal to. The compiler will error if it's wrong.

**Reply by L-as** [reply]
What if you need to, e.g. in a generic function that wants to use `Foo` on an array?

**Reply by eudoxia0** [reply]
I think I understand what you mean. Like this?

```
generic [T: Type(Foo)]
function bar(t: T): Unit is
    let x: ??? := foo(t);
end;
```

**Reply by L-as** [reply]
Yep.

**Reply by eudoxia0** [reply]
Good question :sweat_smile:. I'll look into how Swift does it, they have associated types.
